File: android/push/src/main/kotlin/io/rynbridge/push/PushModule.kt
package io.rynbridge.push

import io.rynbridge.core.*

class PushModule(provider: PushProvider) : BridgeModule {

    override val name = "push"
    override val version = "0.1.0"
    override val actions: Map<String, ActionHandler> = mapOf(
        "register" to { _ ->
            val result = provider.register()
            result.toPayload()
        },
        "unregister" to { _ ->
            provider.unregister()
            emptyMap()
        },
        "getToken" to { _ ->
            val token = provider.getToken()
            mapOf("token" to (token?.let { BridgeValue.string(it) } ?: BridgeValue.nullValue()))
        },
        "requestPermission" to { _ ->
            val granted = provider.requestPermission()
            mapOf("granted" to BridgeValue.bool(granted))
        },
        "getPermissionStatus" to { _ ->
            val result = provider.getPermissionStatus()
            result.toPayload()
        },
        "getInitialNotification" to { _ ->
            val notification = provider.getInitialNotification()
            notification?.toPayload() ?: mapOf(
                "title" to BridgeValue.nullValue(),
                "body" to BridgeValue.nullValue(),
                "data" to BridgeValue.nullValue()
            )
        }
    )
}

File: ios/Sources/RynBridgePush/PushModule.swift
import Foundation
import RynBridge

public struct PushModule: BridgeModule, Sendable {
    public let name = "push"
    public let version = "0.1.0"
    public let actions: [String: ActionHandler]

    public init(provider: PushProvider) {
        actions = [
            "register": { _ in
                let result = try await provider.register()
                return result.toPayload()
            },
            "unregister": { _ in
                try await provider.unregister()
                return [:]
            },
            "getToken": { _ in
                let token = try await provider.getToken()
                return ["token": token.map { .string($0) } ?? .null]
            },
            "requestPermission": { _ in
                let granted = try await provider.requestPermission()
                return ["granted": .bool(granted)]
            },
            "getPermissionStatus": { _ in
                let result = try await provider.getPermissionStatus()
                return result.toPayload()
            },
            "getInitialNotification": { _ in
                let notification = try await provider.getInitialNotification()
                return notification?.toPayload() ?? [
                    "title": .null,
                    "body": .null,
                    "data": .null,
                ]
            },
        ]
    }
}
